db/db_test.cc

```diff
@@ -4054,48 +4054,76 @@ TEST_F(DBTest, ConcurrentFlushWAL) {
-        for (size_t i = 0; i < cnt; i++) {
-          auto istr = ToString(i);
-          db_->Put(wopt, db_->DefaultColumnFamily(), "a" + istr, "b" + istr);
-        }
-      });
-      if (two_write_queues) {
-        threads.emplace_back([&] {
-          for (size_t i = cnt; i < 2 * cnt; i++) {
+  SyncPoint::GetInstance()->SetCallBack("SplitWrite:allowed", [&](void* arg) {
+    size_t& allowed = *(static_cast<size_t**>(arg))[0];
+    size_t size = *(static_cast<size_t**>(arg))[1];
+    // when 'allowed == size', it means we are going to start write the buffer,
+    // then we split write into two parts by allowed /= 2, then first write will
+    // be done immediately, and this callback will be invoked again in the next
+    // time with 'allowed != size', so we slow down the write to increase the
+    // chance of conflict between FlushWAL thread and Writebatch Put thread
+    if (allowed == size && size >= 4) {
+      allowed /= 2;
+    } else {
+      env_->SleepForMicroseconds(1000);  // 1ms
+    }
+  });
+  for (bool splitWriteAndSlowSecondWrite : {true, false}) {
+    if (splitWriteAndSlowSecondWrite) {
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+    }
+    for (bool two_write_queues : {false, true}) {
+      for (bool manual_wal_flush : {false, true}) {
+        for (bool sync : {false, true}) {
+          wopt.sync = sync;
+          options.two_write_queues = two_write_queues;
+          options.manual_wal_flush = manual_wal_flush;
+          options.create_if_missing = true;
+          DestroyAndReopen(options);
+          std::vector<port::Thread> threads;
+          threads.emplace_back([&] {
+            for (size_t i = 0; i < cnt; i++) {
+              auto istr = ToString(i);
+              db_->Put(wopt, db_->DefaultColumnFamily(), "a" + istr,
+                       "b" + istr);
+            }
+          });
+          if (two_write_queues) {
+            threads.emplace_back([&] {
+              for (size_t i = cnt; i < 2 * cnt; i++) {
+                auto istr = ToString(i);
+                WriteBatch batch;
+                batch.Put("a" + istr, "b" + istr);
+                dbfull()->WriteImpl(wopt, &batch, nullptr, nullptr, 0, true);
+              }
+            });
+          }
+          threads.emplace_back([&] {
+            for (size_t i = 0; i < cnt * 100;
+                 i++) {  // FlushWAL is faster than Put
+              db_->FlushWAL(false);
+            }
+          });
+          for (auto& t : threads) {
+            t.join();
+          }
+          options.create_if_missing = false;
+          // Recover from the wal and make sure that it is not corrupted
+          Reopen(options);
+          for (size_t i = 0; i < cnt; i++) {
+            PinnableSlice pval;
             auto istr = ToString(i);
-            WriteBatch batch;
-            batch.Put("a" + istr, "b" + istr);
-            dbfull()->WriteImpl(wopt, &batch, nullptr, nullptr, 0, true);
+            ASSERT_OK(
```

Comment:
We can simply do `ASSERT_EQ("b" + istr, Get("a" + istr))`

Reply:
That diff is code format change.  Now,   add a new ManualFlushWalAndWriteRace test. Current diff does not have this code format change.